### PR TITLE
perf(semantic): eliminate O(N²) host-text rescans in token collection

### PR DESCRIPTION
**Pull request by atusy**
## Summary

Opening a large markdown file made `semanticTokens/full` painfully slow. Flamegraph profiling (`benches/profile/`) showed **58.7% of all self-time inside `collect_host_tokens`**: its prologue recomputed the content-local→host position mapping by scanning the host text from byte 0 (`chars().filter(|c| *c == '\n').count()`) **once per injection**. Markdown turns every paragraph and heading into a `markdown_inline` injection, so the total cost was O(injections × document size) = **O(N²)**.

## Fix

Reuse the line-start byte index that `compute_active_injection_regions` already built to solve the same problem on a smaller scale:

- hoist `build_line_start_bytes` from `parallel.rs` into `token_collector.rs`
- build it **once per request** (`handle_semantic_tokens_full` / `collect_injection_tokens_parallel`)
- thread `host_line_starts: &[usize]` through `process_injection_sync` → `collect_host_tokens`, replacing the O(document) scan with a `partition_point` binary search

Also adds `--file` to `benches/profile/drive.py` so a real document (not just generated ones) can be replayed through the synchronous driver.

## Results (profiling build, generated markdown, ms/request)

| code blocks | before | after | speedup |
|---|---|---|---|
| 500 | 72 | 26 | 2.8× |
| 1000 | 505 | 58 | 8.8× |
| 2000 | 1,570 | 141 | 11× |
| 4000 | 5,969 | 391 | 15× |

After the fix `collect_host_tokens` no longer appears in the flamegraph self-time top 10; the remaining compute hotspot is `finalize_tokens` (~13%).

## Notes

- No behavioral change intended: positions are computed identically (covered by the existing 1,571 lib tests; the line/col math edge cases — offset at a newline byte, offset 0, multibyte — behave the same under both formulations).
- The original 7-second user report was amplified by running a debug build with `RUST_LOG=trace`; the quadratic curve above is measured on the optimized build and is real independent of that.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: benches/profile/drive.py [hunk 1]
```diff
@@ -36,7 +38,11 @@ def main() -> None:
     if args.requests <= 0:
         ap.error("--requests must be positive")  # avoids divide-by-zero in the summary
 
-    if args.lang == "rust":
+    if args.file:
+        ext = "rs" if args.lang == "rust" else "md"
+        with open(args.file) as f:
```
[the remark this answers is not in the document]
Comment (atusy) [reply benches/profile/drive.py]: Fixed in 68f7b1e2 — `open(args.file, encoding="utf-8")`. Thanks!

File: src/analysis/semantic/parallel.rs [hunk 2]
```diff
@@ -441,6 +447,10 @@ pub(crate) fn collect_injection_tokens_parallel(
     // Create factory from coordinator's registry (cloned for thread safety)
     let factory = ThreadLocalParserFactory::new(coordinator.language_registry_for_parallel());
 
+    // Built once and shared by every injection (and the active-region conversion
+    // below) so byte→line/col mapping never rescans the host text per injection.
+    let host_line_starts = build_line_start_bytes(host_text);
+
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/parallel.rs]: Fixed in 0705a775 — `collect_injection_tokens_parallel` now takes `host_line_starts: &[usize]` and `handle_semantic_tokens_full` passes the index it already built, so the host text is scanned once per request. Thanks!

File: benches/profile/drive.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (atusy) [reply benches/profile/drive.py]: Fixed in 68f7b1e2 with an explicit `encoding="utf-8"`. Thanks!

File: benches/profile/drive.py [hunk 3]
```diff
@@ -36,7 +38,11 @@ def main() -> None:
     if args.requests <= 0:
         ap.error("--requests must be positive")  # avoids divide-by-zero in the summary
 
-    if args.lang == "rust":
+    if args.file:
+        ext = "rs" if args.lang == "rust" else "md"
+        with open(args.file, encoding="utf-8") as f:
+            uri, lang, text = f"file:///profile/input.{ext}", args.lang, f.read()
```
[the remark this answers is not in the document]
Comment (atusy) [reply benches/profile/drive.py]: Applied your suggestion in d2b334cc — language and URI extension are inferred from the file extension (.md/.rs), with --lang as the fallback for unrecognized extensions. The summary line now reports the effective language too. Thanks!

File: src/analysis/semantic/parallel.rs [hunk 4]
```diff
@@ -546,23 +557,6 @@ fn compute_active_injection_regions(
...
-        if b == b'\n' {
-            starts.push(i + 1);
-        }
-    }
-    starts
-}
-
 /// Convert a byte offset in host_text to a (line, utf16_col) pair.
 ///
 /// `line_starts` must come from [`build_line_start_bytes`] for the same text.
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/parallel.rs]: Fixed in d2b334cc — the link now uses the module path (`super::token_collector::build_line_start_bytes`). Verified no warning with `cargo doc --document-private-items`. Thanks!

File: benches/profile/drive.py [hunk 5]
```diff
@@ -129,8 +135,11 @@ def read_until(want_id):
             srv.kill()
             srv.wait()
 
+    n_lines = text.count("\n")
```
[the remark this answers is not in the document]
Comment (atusy) [reply benches/profile/drive.py]: Fixed in d2b334cc — `len(text.splitlines())` now counts the final unterminated line too (verified: a no-trailing-newline file reports 1L). Thanks!
